mongo/src/main/java/com/redhat/lightblue/mongo/crud/Utils.java

```diff
@@ -0,0 +1,71 @@
+/*
+ Copyright 2013 Red Hat, Inc. and/or its affiliates.
+
+ This file is part of lightblue.
+
+ This program is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.redhat.lightblue.mongo.crud;
+
+import java.util.Set;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.ArrayList;
+
+import com.mongodb.DBCollection;
+import com.mongodb.DBCursor;
+import com.mongodb.BasicDBObject;
+
+public final class Utils {
+
+    /**
+     * Returns the _ids of documents that are failed to be updated
+     *
+     * @param collection The DB collection
+     * @param currentTxId The txid of the current operation. The
+     * docversion field of all docs are set to this value for the
+     * current operation
+     * @param idToTxIdMap A mapping of doc _id to the old transaction
+     * id of the document
+     *
+     * @return A set of _ids whose `document version is not equal to currentTxId
+     */
+    public static Set<Object> checkFailedUpdates(DBCollection collection,
+                                                 String currentTxId,
+                                                 Map<Object,String> idToTxIdMap) {
+        // We would like to find all _id:currentTxId documents. Those
+        // that are missing, i.e. not in the resultset, are the docs
+        // that are modified during or after our update operation
+        DBCursor cursor=null;
+        try {
+            Set<Object> keys=idToTxIdMap.keySet();
+            cursor=collection.find(new BasicDBObject(Translator.DOC_VERSION_FULLPATH_STR,currentTxId).
```

Comment:
I'm not sure how to provide this feature otherwise, but I think this has a race condition, still, where some updates could be counted as failed even though they actually succeeded. This would happen if, before this find request, another update happened which changed the document version of one of the documents involved in this update. So, I think even though both updates were successful, the earlier one would detect a doc version change since the start of the update transaction, and fail that request.

It's a shame BulkWriteResult doesn't contain more useful info! Is there any way to get the ids of matched documents back from mongo that is still as efficient as BWO?